File: src/java/tusan/nodes/Statement.java
package net.tutla.tums.tusan.nodes;

import net.tutla.tums.tusan.Node;
import net.tutla.tums.tusan.lexer.Token;
import net.tutla.tums.tusan.lexer.TokenType;
import net.tutla.tums.tusan.nodes.base.If;
import net.tutla.tums.tusan.nodes.base.Return;
import net.tutla.tums.tusan.nodes.base.function.FunctionNode;
import net.tutla.tums.tusan.nodes.base.loops.Loop;
import net.tutla.tums.tusan.nodes.base.loops.While;
import net.tutla.tums.tusan.nodes.expression.Expression;
import net.tutla.tums.tusan.nodes.tums.On;

import java.util.Arrays;
import java.util.List;

public class Statement extends Node {
    private final List<TokenType> validTypes = Arrays.asList(TokenType.STRUCTURE, TokenType.KEYWORD, TokenType.IDENTIFIER, TokenType.EFFECT, TokenType.LEFT_CURLY, TokenType.NUMBER, TokenType.STRING);
    public Statement(Token token){
        super(token);
    }
    public Object create(){

        Object value = false;
        if (interpreter.end){
            return false;
        }

        if (token.type == TokenType.LEFT_CURLY){
            new Statement(interpreter.nextToken()).create();
            interpreter.expectTokenType(TokenType.RIGHT_CURLY);
            return value;
        }
        if (validTypes.contains(token.type)){
            if (token.type == TokenType.EFFECT){
                value = new Effect(token).create();
            } else if (token.type == TokenType.STRUCTURE){
                if (token.value.equals("if")){
                    new If(interpreter.nextToken()).create();
                } else if (token.value.equals("while")) {
                    new While(interpreter.nextToken()).create();
                } else if (token.value.equals("loop")) {
                    new Loop(interpreter.nextToken()).create();
                } else if (token.value.equals("on")) {
                    new On(token).create();
                } else if (token.value.equals("function")) {
                    new FunctionNode(interpreter.nextToken()).create();
                } else {
                    interpreter.error("UnexpectedToken", "Structure "+token.type.name()+":"+token.value+" has no definition", null);
                }
            } else if (Arrays.asList(TokenType.IDENTIFIER, TokenType.LEFT_CURLY, TokenType.NUMBER, TokenType.STRING, TokenType.KEYWORD).contains(token.type)){
                value = new Expression(token).create();
            } else if (token.type ==  TokenType.BREAKSTRUCTURE){
                value = new Return(token).create();
            } else {
                interpreter.error("UnexpectedToken", "Expected valid statement got "+token.type.name()+":"+token.value, null);
            }
        } else if (token.type == TokenType.BREAKSTRUCTURE && token.value.equals("return")){
            new Return(token).create();
        } else {
            interpreter.error("UnexpectedToken", "Expected valid statement got "+token.type.name()+":"+token.value, null);
        }
        return value;
    }
}

File: src/python/src/tusan/nodes/Statement.py
import time, asyncio

from tusan.Node import Node
from tusan.lexer.Token import Token
from tusan.Variable import  Variable
from tusan.nodes.base.If import *
from tusan.nodes.base.function import *
from tusan.nodes.base.loops import *
from tusan.nodes.expression import *
from tusan.nodes.base.Return import ReturnNode
from tusan.nodes.effects import *
class StatementNode(Node):
    def __init__(self, token:Token):
        self.interpreter = token.interpreter
        self.auto_eval = True
        self.token = token

    async def create(self):
        self.interpreter.debug_msg(self.token, "<- stmt (node) token")
        if self.interpreter.end_found:
            return False
        if self.token.type in ["KEYWORD", "IDENTIFIER","STRUCTURE","EFFECT","LEFT_CURLY","NUMBER","STRING"]:
            if self.token.type == "EFFECT":
                await EffectNode(self.token).create()
            elif self.token.type == "STRUCTURE":
                if self.token.value == "if":
                    await IfNode(self.interpreter.next_token()).create()
                elif self.token.value == "function":
                    await FunctionNode(self.interpreter.next_token()).create()
                elif self.token.value == "while":
                    await WhileNode(self.interpreter.next_token()).create()
                elif self.token.value == "loop":
                    await LoopNode(self.interpreter.next_token()).create()
                    
                elif self.token.value == "on":
                    from tusan.nodes.discord.base.on import OnNode
                    await OnNode(self.token).create()
            elif self.token.type == "IDENTIFIER":
                await ExpressionNode(self.token).create()
            elif self.token.type == "KEYWORD":
                await ExpressionNode(self.token).create()
            elif self.token.type in ["LEFT_CURLY","NUMBER","STRING"]:
                await ExpressionNode(self.token).create()
            elif self.token.type == "BREAKSTRUCTURE":
                await ReturnNode(self.token).create()
            else: self.interpreter.error("UnexpectedToken", f"Expected KEYWORD | VALID_IDENTIFIER | STRUCTURE, got {self.interpreter.current_token.type} @tusan {self.interpreter.current_token.value}{self.token}", notes=["Possible Fix: Recheck code with documentation, you might have missed a keyword at position"])

        else: self.interpreter.error("UnexpectedToken", f"Expected KEYWORD | VALID_IDENTIFIER | STRUCTURE, got {self.interpreter.current_token.type} @tusan {self.interpreter.current_token.value}{self.token}", notes=["Possible Fix: Recheck code with documentation, you might have missed a keyword at position"])

        self.type="1en"
        self.interpreter.debug_msg(self.interpreter.current_token, "<- stmt (node) end\n\n")
        return self